application/src/main/java/org/togetherjava/tjbot/commands/moderation/WarnCommand.java

```diff
@@ -76,45 +97,38 @@ public WarnCommand(@NotNull ModerationActionsStore actionsStore) {
                 target, dmNoticeText, reason);
     }
 
-    private boolean handleChecks(@NotNull Member bot, @NotNull Member author,
-            @NotNull CharSequence reason, @NotNull Guild guild, @NotNull Interaction event) {
-
-        if (!ModerationUtils.handleHasAuthorRole(ACTION_VERB, hasRequiredRole, author, event)) {
-            return false;
-        }
-        if (!ModerationUtils.handleHasBotPermissions(ACTION_VERB, Permission.BAN_MEMBERS, bot,
-                guild, event)) {
-            return false;
-        }
-        if (!ModerationUtils.handleHasAuthorPermissions(ACTION_VERB, Permission.BAN_MEMBERS, author,
-                guild, event)) {
-            return false;
-        }
-        return ModerationUtils.handleReason(reason, event);
-    }
-
     @Override
     public void onSlashCommand(@NotNull SlashCommandEvent event) {
-        OptionMapping userOption =
-                Objects.requireNonNull(event.getOption(USER_OPTION), "The option is null");
-        User target = userOption.getAsUser();
+        OptionMapping targetOption =
+                Objects.requireNonNull(event.getOption(USER_OPTION), "The target is null");
+        User target = targetOption.getAsUser();
         Member author = Objects.requireNonNull(event.getMember(), "The author is null");
         Guild guild = Objects.requireNonNull(event.getGuild(), "The guild is null");
         String reason = Objects.requireNonNull(event.getOption(REASON_OPTION), "The reason is null")
             .getAsString();
 
-        Member bot = guild.getSelfMember();
-
-        if (!handleChecks(bot, author, reason, guild, event)) {
+        if (!handleChecks(guild.getSelfMember(), author, targetOption.getAsMember(), reason,
+                event)) {
             return;
         }
-        long userId = target.getIdLong();
-        dmUser(userId, reason, event)
-            .map(hasSentDm -> sendFeedback(hasSentDm, target, author, reason))
-            .flatMap(event::replyEmbeds)
-            .queue();
 
-        actionsStore.addAction(guild.getIdLong(), author.getIdLong(), target.getIdLong(),
-                ModerationUtils.Action.WARN, null, reason);
+        warnUserFlow(target, author, reason, guild, event).queue();
+    }
+
+    @SuppressWarnings("BooleanMethodNameMustStartWithQuestion")
+    private boolean handleChecks(@NotNull Member bot, @NotNull Member author,
+            @Nullable Member target, String reason, @NotNull SlashCommandEvent event) {
+        if (target == null) {
+            if (!ModerationUtils.handleHasAuthorRole(ACTION_VERB, hasRequiredRole, author, event)) {
+                return false;
+            }
+        } else {
+            // Check if the author can warn a user with higher perms.
+            if (!ModerationUtils.handleCanInteractWithTarget(ACTION_VERB, bot, author, target,
+                    event)) {
+                return false;
+            }
+        }
+        return ModerationUtils.handleReason(reason, event);
```

Comment:
This is wrong. You have to execute the first check with the role in all cases, not just if its null. Otherwise this enables **everyone** to warn people who left the guild.